### PR TITLE
[opentitantool] Improve UART on ti50 emulation

### PR DESCRIPTION
**Pull request by jesultra**
Host emulation works by compiling the GSC code into a Linux executable, instead of an image for OpenTitan, and running this executable as a sub-process to `opentitansession`.

The sub-process will create a number of unix pipes, for UARTs and other peripherals, and will use these to communicate with the simulation environment.

Prior to this CL, `opentitantool console` would not capture the first serial console output, because of a race.  The sub-procces would create the unix pipes, tell opentitansession that it was ready, and then immediately proceed to produce output to the unix pipe representing the serial console.  This allowed opentitansession no time to connect to the pipe, and any data sent before a reading process had connected would be lost.

This CL introduces a handshake, such that the sub-process is instructed to wait after having indicated that it is done creating Unix pipes. Also, opentitansession will now connect to all UART pipes on sub-process startup, rather than only at first read operation.  Together, these changes allow the emulator to guarantee that the `opentitantool console` will show the very first serial console output as the emulated chip comes out of reset.

### CODE REVIEW

File: sw/host/opentitanlib/src/transport/ti50emulator/uart.rs [hunk 1]
```diff
@@ -56,15 +50,14 @@ impl Ti50Uart {
...
-        self.reconnect()?;
         return Ok(RefMut::map(self.socket.borrow_mut(), |socket| {
             socket.as_mut().unwrap()
         }));
     }
 }
 
 /// A trait which represents a UART.
-impl Uart for Ti50Uart {
+impl Uart for Rc<Ti50Uart> {
```
Comment (jesultra) [inline sw/host/opentitanlib/src/transport/ti50emulator/uart.rs:60]: In this CL, I wanted to have `Rc<dyn Uart>` and `Rc<Ti50Uart>` simutaneously referring to the same object.  (The former for returning from the `Transport` trait, and the latter for allowing logic internal to the transport relating to starting a new sub-process to iterate over Uart ports, and call an internal method on each.

The way I found to allow the two kinds of Rc's was to not have the `Uart` trait implemented directly by `Ti50Uart`, but by the wrapped `Rc<Ti50Uart>`.  Apparently, Rust implicitly deferences all the mentions of `self` in the implementing code.  I could want to use a similar idiom in other transport backends, which may want to have internal methods, accessible from elsewhere within the transport implementation.

File: sw/host/opentitanlib/src/transport/ti50emulator/uart.rs [hunk 2]
```diff
@@ -56,15 +50,14 @@ impl Ti50Uart {
     }
 
     pub fn get_socket(&self) -> Result<RefMut<UnixStream>> {
-        self.reconnect()?;
```
Comment (semihalf-mazurek-michal) [inline sw/host/opentitanlib/src/transport/ti50emulator/uart.rs:59]: I see that you are moving this to spawn_process(...)
What command OTT was used to tested this changes.

**Review by semihalf-mazurek-michal — COMMENTED** [review]
Could it be implemented as a separate socket?

File: sw/host/opentitanlib/src/transport/ti50emulator/uart.rs (hunk 1 above)
Comment (semihalf-mazurek-michal) [reply sw/host/opentitanlib/src/transport/ti50emulator/uart.rs:60]: @cfrantz 
Can you look at that Rust language construction ?

File: sw/host/opentitanlib/src/transport/ti50emulator/mod.rs [hunk 3]
```diff
@@ -91,8 +93,10 @@ impl Ti50Emulator {
             gpio_map.insert(name.to_uppercase(), Rc::clone(&gpio));
         }
         for (name, path) in conf.uart.iter() {
-            let uart: Rc<dyn Uart> = Rc::new(Ti50Uart::open(&inner, path)?);
-            uart_map.insert(name.to_uppercase(), Rc::clone(&uart));
+            let uart: Rc<Ti50Uart> = Rc::new(Ti50Uart::open(&inner, path)?);
```
[the remark this answers is not in the document]
Comment (jesultra) [reply sw/host/opentitanlib/src/transport/ti50emulator/mod.rs:96]: As the PR is, UART data goes through unix named pipes, while the handshake/control goes through standard input/output.  The reason why the code uses both, is that at the time of the control handshake, is when we want the parent process to establish connections to each of the named pipes used for UART data.

File: sw/host/opentitanlib/src/transport/ti50emulator/uart.rs (hunk 2 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/transport/ti50emulator/uart.rs:59]: I tested this by this sequence of commands (from within ti50/common), first:
```
$ cp build/system_test_auto/host_emulation/host_emulation/image.A.bin /tmp/system_test_auto.bin && /mnt/host/source/src/third_party/lowrisc/opentitan/bazel-bin/sw/host/opentitansession/opentitansession --foreground --rcfile= --logging=warn --conf ports/host_emulation/software/tools/host_emulation.json --conf ports/host_emulation/software/tools/ti50_host_emulation.json --interface ti50emulator --executable-directory /tmp/ --executable system_test_auto.bin
```
Then in a separate window, I did:
```
$ /mnt/host/source/src/third_party/lowrisc/opentitan/bazel-bin/sw/host/opentitantool/opentitantool --rcfile= --format=json --interface=proxy --proxy=127.0.0.1 --port 9900 console -q --uart console
```
Finally, I released the emulator by:
```
$ /mnt/host/source/src/third_party/lowrisc/opentitan/bazel-bin/sw/host/opentitantool/opentitantool --rcfile= --format=json --interface=proxy --proxy=127.0.0.1 --port 9900 gpio write RESET true
```

What I noted was that with the PR, the console in the second command above would show me all output, starting from `Starting RO 1.2.3 RW 1.2.3`.  Without this PR, a lot of the initial output is omitted, causing tests to not be able to reliably recognize e.g. `##KERNEL TESTS START` in the output.

I also tried setting the `RESET` to `false` and `true` again multiple times, each time seeing output from the new sub-process.

**Comment by jesultra** [comment]
> Could it be implemented as a separate socket?

I am not sure what the point would be.  As is, UART data goes through four unix named pipes, and the handshake is on stdin/stdout.  No actual serial data goes through stdin/stdout in the way that opentitansession uses host emulation.  (There is another mode of the ti50 emulation in which three UARTs use unix named pipes, and the fourth uses stdio/stdout, but that is not used by opentitansession.)

> If we separate UART and Control socket we could avoid this misleading tricks [ `Rc<Rc<...>>` ].

I am not so sure about that.  The multi-level `Rc<>` is my workaround for having `Rc<dyn Uart>` and `Rc<Ti50Uart>` simultaneously referring to the same `Ti50Uart` struct instance.  This is required such that the handshake logic can iterate over the emulator UARTs, and call a non-public method (`connect()`) on each of them.  That need would still exist no matter what kind of socket was used for the handshake.

File: sw/host/opentitanlib/src/transport/ti50emulator/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (semihalf-mazurek-michal) [reply sw/host/opentitanlib/src/transport/ti50emulator/mod.rs:96]: In that case I withdraw the question, STD I/O should be ok.

File: sw/host/opentitanlib/src/transport/ti50emulator/uart.rs (hunk 2 above)
Comment (semihalf-mazurek-michal) [reply sw/host/opentitanlib/src/transport/ti50emulator/uart.rs:59]: Thanks i will check the result on my side and let you know the outcome.

Comment (semihalf-mazurek-michal) [reply sw/host/opentitanlib/src/transport/ti50emulator/uart.rs:59]: LGTM

**Review by semihalf-mazurek-michal — APPROVED** [review]
I checked the handshake code and it is LTGM.

Please consult Rust construction with `Rc<Ti50Uart>` with other reviewers.
Maybe someone will come with better solution to this problem.

File: sw/host/opentitanlib/src/transport/ti50emulator/uart.rs (hunk 1 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/transport/ti50emulator/uart.rs]: Chris has explained to me that contrary to my previous belief and statements, you can in fact have `Rc<dyn Uart>` and `Rc<Ti50Uart>` simultaneously referring to the same struct.  I have updated the code accordingly, avoiding implementing a trait on `Rc<Ti50Uart>`.

File: sw/host/opentitanlib/src/transport/ti50emulator/emu.rs [hunk 4]
```diff
@@ -283,6 +295,16 @@ impl EmulatorProcess {
...
                 if PATTERN.as_bytes() == &buffer[..newline_pos] {
+                    // Connect to Unix pipes representing each UART.
+                    for weak_rc_uart in &*inner.uarts.borrow() {
+                        match weak_rc_uart.upgrade() {
+                            None => (),
+                            Some(uart) => uart.connect()?,
+                        }
+                    }
+                    // Now release the sub-process to run project code.
+                    child_stdin.write_all("GO\n".as_bytes())?;
```
Comment (cfrantz) [inline sw/host/opentitanlib/src/transport/ti50emulator/emu.rs:306]: You can probably just write `b"Go\n"` here.

File: sw/host/opentitanlib/src/transport/ti50emulator/mod.rs [hunk 5]
```diff
@@ -74,6 +75,7 @@ impl Ti50Emulator {
             instance_directory,
             process: RefCell::new(process),
             gpio_map: RefCell::new(HashMap::new()),
+            uarts: RefCell::new(Vec::new()),
```
Comment (cfrantz) [inline sw/host/opentitanlib/src/transport/ti50emulator/mod.rs:78]: IIRC, the Default impl for containers is an empty container, and I think `RefCell<T>` implements `Default` if `T: Default`.

Depending on your sense of style, you could write `RefCell::default()` here. (I'm find if you want to leave it as-is also).

**Review by cfrantz — APPROVED** [review]

File: sw/host/opentitanlib/src/transport/ti50emulator/emu.rs (hunk 4 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/transport/ti50emulator/emu.rs:306]: Done

File: sw/host/opentitanlib/src/transport/ti50emulator/mod.rs (hunk 5 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/transport/ti50emulator/mod.rs:78]: This is good to know.  Done.

**Pull request merged**